cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -316,5 +318,165 @@ TEST_F(TestIfElseKernel, IfElseDispatchBest) {
+  CheckVarArgs("case_when", {cond1, values1, cond2, values2},
+               ArrayFromJSON(type, builder.str()));
+  // Else clauses/mixed scalar and array
+  builder.str("");
+  builder << "[null, " << value1 << ',' << value1 << ',' << value2 << ']';
+  CheckVarArgs("case_when", {cond1, values1, cond2, values2, scalar1},
+               ArrayFromJSON(type, builder.str()));
+  CheckVarArgs("case_when", {cond1, values1, cond2, values2, values1},
+               ArrayFromJSON(type, builder.str()));
+}
+
+TYPED_TEST(TestCaseWhenNumeric, FixedSize) {
+  auto type = default_type_instance<TypeParam>();
+  CheckCaseWhenCases(type, "10", "42");
+}
+
+TEST(TestCaseWhen, Null) {
+  auto scalar = ScalarFromJSON(null(), "null");
+  auto array = ArrayFromJSON(null(), "[null, null, null, null]");
+  CheckVarArgs("case_when", {array}, array);
+  CheckVarArgs("case_when", {scalar, array}, array);
+  CheckVarArgs("case_when", {scalar, array, array}, array);
+}
+
+TEST(TestCaseWhen, Boolean) { CheckCaseWhenCases(boolean(), "true", "false"); }
+
+TEST(TestCaseWhen, DayTimeInterval) {
+  CheckCaseWhenCases(day_time_interval(), "[10, 2]", "[2, 5]");
+}
+
+TEST(TestCaseWhen, Decimal) {
+  for (const auto& type :
+       std::vector<std::shared_ptr<DataType>>{decimal128(3, 2), decimal256(3, 2)}) {
+    CheckCaseWhenCases(type, "\"1.23\"", "\"4.56\"");
+  }
+}
+
+TEST(TestCaseWhen, FixedSizeBinary) {
+  auto type = fixed_size_binary(3);
+  CheckCaseWhenCases(type, "\"aaa\"", "\"bbb\"");
+}
+
+TEST(TestCaseWhen, DispatchBest) {
+  auto Check = [](std::vector<ValueDescr> original_values,
+                  std::vector<ValueDescr> expected_equivalent_values) {
+    EXPECT_OK_AND_ASSIGN(auto function, GetFunctionRegistry()->GetFunction("case_when"));
+    auto values = original_values;
+    ARROW_ASSIGN_OR_RAISE(auto actual_kernel, function->DispatchBest(&values));
+    EXPECT_OK_AND_ASSIGN(auto expected_kernel,
+                         function->DispatchBest(&expected_equivalent_values));
+    EXPECT_EQ(actual_kernel, expected_kernel)
+        << "  DispatchBest" << ValueDescr::ToString(original_values) << " => "
+        << actual_kernel->signature->ToString() << "\n"
+        << "  DispatchBest" << ValueDescr::ToString(expected_equivalent_values) << " => "
+        << expected_kernel->signature->ToString();
+    return Status::OK();
+  };
+
+  ASSERT_OK(Check({int32()}, {int32()}));
+  ASSERT_OK(Check({boolean(), int32(), int32()}, {boolean(), int32(), int32()}));
+  ASSERT_OK(Check({null(), int32(), int32()}, {boolean(), int32(), int32()}));
+  ASSERT_OK(Check({boolean(), int32(), int8()}, {boolean(), int32(), int32()}));
+  ASSERT_OK(Check({boolean(), int32(), uint32()}, {boolean(), int64(), int64()}));
+  ASSERT_RAISES(Invalid,
+                Check({boolean(), utf8(), int32()}, {boolean(), int32(), int32()}));
+  ASSERT_RAISES(Invalid,
+                Check({int32(), int32(), int32()}, {boolean(), int32(), int32()}));
+}
```

Comment:
Somewhere, you should also test what happens when `case_when` is called with zero arguments.